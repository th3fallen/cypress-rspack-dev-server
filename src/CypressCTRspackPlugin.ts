import type { Compilation, Compiler } from '@rspack/core'
import type { EventEmitter } from 'events'
import isEqual from 'lodash/isEqual'
import fs, { PathLike } from 'fs-extra'
import path from 'path'
import debugLib from 'debug'

const debug = debugLib('cypress-rspack-dev-server:rspackPlugin')

type UtimesSync = (
  path: PathLike,
  atime: string | number | Date,
  mtime: string | number | Date,
) => void

export interface CypressCTRspackPluginOptions {
  files: Cypress.Cypress['spec'][]
  projectRoot: string
  supportFile: string | false
  devServerEvents: EventEmitter
  indexHtmlFile: string
}

export type CypressCTContextOptions = Omit<
  CypressCTRspackPluginOptions,
  'devServerEvents' | 'rspack'
>

export interface CypressCTRspackContext {
  _cypress: CypressCTContextOptions
}

/**
 * @internal
 */
export type RspackCompilation = Compilation & {
  inputFileSystem: {
    fileSystem: {
      utimesSync: UtimesSync
    }
  }
}

export const normalizeError = (error: Error | string) => {
  return typeof error === 'string' ? error : error.message
}

/**
 * A rspack compatible Cypress Component Testing Plugin
 *
 * @internal
 */
export class CypressCTRspackPlugin {
  private files: Cypress.Cypress['spec'][] = []
  private supportFile: string | false
  private compilation: RspackCompilation | null = null
  private indexHtmlFile: string

  private readonly projectRoot: string
  private readonly devServerEvents: EventEmitter

  constructor(options: CypressCTRspackPluginOptions) {
    this.files = options.files
    this.supportFile = options.supportFile
    this.projectRoot = options.projectRoot
    this.devServerEvents = options.devServerEvents
    this.indexHtmlFile = options.indexHtmlFile
  }

  private addLoaderContext = (loaderContext: object) => {
    ;(loaderContext as CypressCTRspackContext)._cypress = {
      files: this.files,
      projectRoot: this.projectRoot,
      supportFile: this.supportFile,
      indexHtmlFile: this.indexHtmlFile,
    }
  }

  private beforeCompile = async (compilationParams: object, callback: Function) => {
    if (!this.compilation) {
      callback()

      return
    }

    // Ensure we don't try to load files that have been removed from the file system
    // but have not yet been detected by the onSpecsChange handler

    const foundFiles = await Promise.all(
      this.files.map(async (file) => {
        try {
          const exists = await fs.pathExists(file.absolute)

          return exists ? file : null
        } catch (e) {
          return null
        }
      }),
    )

    this.files = foundFiles.filter((file) => file !== null) as Cypress.Spec[]

    callback()
  }

  /*
   * `rspack --watch` watches the existing specs and their dependencies for changes.
   * When new specs are created, we need to trigger a recompilation to add the new specs
   * as dependencies. This hook informs rspack that `component-index.html` has been "updated on disk",
   * causing a recompilation (and pulling the new specs in as dependencies). We use the component
   * index file because we know that it will be there since the project is using Component Testing.
   *
   * We were using `browser.js` before to cause a recompilation but we ran into an
   * issue with MacOS Ventura that will not allow us to write to files inside of our application bundle.
   *
   * See https://github.com/cypress-io/cypress/issues/24398
   */
  private onSpecsChange = async (specs: Cypress.Cypress['spec'][]) => {
    if (!this.compilation || isEqual(specs, this.files)) {
      return
    }

    this.files = specs
    const inputFileSystem = this.compilation.inputFileSystem
    // TODO: don't use a sync fs method here
    const utimesSync: UtimesSync = inputFileSystem.fileSystem.utimesSync ?? fs.utimesSync

    utimesSync(path.join(this.projectRoot, this.indexHtmlFile), new Date(), new Date())
  }

  /**
   * The rspack compiler generates a new `compilation` each time it compiles, so
   * we have to apply hooks to it fresh each time
   *
   * @param compilation `RspackCompilation`
   */
  private addCompilationHooks = (compilation: RspackCompilation) => {
    this.compilation = compilation

    // still use legacy `webpack` here since in version 0.x.x does not have rspack
    const compiler = (
      'rspack' in this.compilation.compiler
        ? this.compilation.compiler.rspack
        : (this.compilation.compiler as Compiler).webpack
    ) as Compiler['webpack']

    const loader = compiler.NormalModule.getCompilationHooks(compilation).loader

    loader.tap('CypressCTPlugin', this.addLoaderContext)
  }

  /**
   * The plugin's entrypoint, called once by rspack when the compiler is initialized.
   */
  apply(compiler: unknown): void {
    const _compiler = compiler as Compiler

    this.devServerEvents.on('dev-server:specs:changed', this.onSpecsChange)
    _compiler.hooks.beforeCompile.tapAsync('CypressCTPlugin', this.beforeCompile)
    _compiler.hooks.compilation.tap('CypressCTPlugin', (compilation) =>
      this.addCompilationHooks(compilation as RspackCompilation),
    )
    _compiler.hooks.done.tap('CypressCTPlugin', () => {
      this.devServerEvents.emit('dev-server:compile:success')
    })
  }
}
