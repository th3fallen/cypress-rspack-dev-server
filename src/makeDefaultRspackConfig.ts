import path from 'path'
import debugLib from 'debug'
import { HtmlRspackPlugin, type Configuration } from '@rspack/core'
import type { CreateFinalRspackConfig } from './createRspackDevServer'
import { CypressCTRspackPlugin } from './CypressCTRspackPlugin'

const debug = debugLib('cypress-rspack-dev-server:makeDefaultRspackConfig')

const OUTPUT_PATH = path.join(__dirname, 'dist')

const OsSeparatorRE = RegExp(`\\${path.sep}`, 'g')
const posixSeparator = '/'

export function makeCypressRspackConfig(config: CreateFinalRspackConfig): Configuration {
  const {
    devServerConfig: {
      cypressConfig: {
        experimentalJustInTimeCompile,
        projectRoot,
        devServerPublicPathRoute,
        supportFile,
        indexHtmlFile,
        isTextTerminal: isRunMode,
      },
      specs: files,
      devServerEvents,
    },
  } = config

  const optimization: Record<string, any> = {
    // To prevent files from being tree shaken by rspack, we set optimization.sideEffects: false ensuring that
    // rspack does not recognize the sideEffects flag in the package.json and thus files are not unintentionally
    // dropped during testing in production mode.
    sideEffects: false,
    splitChunks: { chunks: 'all' },
  }

  const publicPath =
    path.sep === posixSeparator
      ? path.join(devServerPublicPathRoute, posixSeparator)
      : // The second line here replaces backslashes on windows with posix compatible slash
        // See https://github.com/cypress-io/cypress/issues/16097
        path.join(devServerPublicPathRoute, posixSeparator).replace(OsSeparatorRE, posixSeparator)

  const finalConfig: Configuration = {
    mode: 'development',
    optimization,
    output: {
      filename: '[name].js',
      path: OUTPUT_PATH,
      publicPath,
    },
    plugins: [
      new HtmlRspackPlugin({
        template: indexHtmlFile,
        filename: 'index.html',
      }),
      new CypressCTRspackPlugin({
        files,
        projectRoot,
        devServerEvents,
        supportFile,
        indexHtmlFile,
      }),
    ],
    devtool: 'inline-source-map',
  }

  if (isRunMode) {
    // if experimentalJustInTimeCompile is configured, we need to watch for file changes as the spec entries are going to be updated per test
    const ignored = experimentalJustInTimeCompile ? /node_modules/ : '**/*'

    // Disable file watching when executing tests in `run` mode
    finalConfig.watchOptions = { ignored }
  }

  return finalConfig
}
