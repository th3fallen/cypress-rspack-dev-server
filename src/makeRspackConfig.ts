import { debug as debugFn } from 'debug'
import * as path from 'path'
import { merge } from 'webpack-merge'
import { importModule } from 'local-pkg'
import type { Configuration, EntryObject } from '@rspack/core'
import { makeCypressRspackConfig } from './makeDefaultRspackConfig'
import type { CreateFinalRspackConfig } from './createRspackDevServer'
import { configFiles } from './constants'
import { dynamicImport } from './dynamic-import'

const debug = debugFn('cypress-rspack-dev-server:makeRspackConfig')

const removeList = [
  // We provide a webpack-html-plugin config pinned to a specific version (4.x)
  // that we have tested and are confident works with all common configurations.
  // https://github.com/cypress-io/cypress/issues/15865
  'HtmlWebpackPlugin',

  // the rspack's internal html plugin
  'HtmlRspackPlugin',

  // We already reload when rspack recompiles (via listeners on
  // devServerEvents). Removing this plugin can prevent double-refreshes
  // in some setups.
  'HotModuleReplacementPlugin',
]

export const CYPRESS_RSPACK_ENTRYPOINT = path.resolve(__dirname, 'browser.js')

/**
 * Removes and/or modifies certain plugins known to conflict
 * when used with @rspack/dev-server.
 */
function modifyRspackConfigForCypress(rspackConfig: Partial<Configuration>) {
  if (rspackConfig?.plugins) {
    rspackConfig.plugins = rspackConfig.plugins.filter(
      (plugin) => plugin && !removeList.includes(plugin.constructor.name),
    )
  }

  delete rspackConfig.entry
  delete rspackConfig.output

  return rspackConfig
}

async function getRspackConfigFromProjectRoot(projectRoot: string) {
  const { findUp } = await dynamicImport<typeof import('find-up')>('find-up')

  return await findUp(configFiles, { cwd: projectRoot })
}

/**
 * Creates a rspack compatible rspack "configuration" to pass to the sourced rspack function
 */
export async function makeRspackConfig(config: CreateFinalRspackConfig): Promise<Configuration> {
  let userRspackConfig = config.devServerConfig.rspackConfig
  const frameworkRspackConfig = config.frameworkConfig as Partial<Configuration>
  const {
    cypressConfig: { projectRoot, supportFile },
    specs: files,
    framework,
  } = config.devServerConfig

  if (!userRspackConfig && !frameworkRspackConfig) {
    debug('Not user or framework rspack config received. Trying to automatically source it')

    const configFile = await getRspackConfigFromProjectRoot(projectRoot)

    if (configFile) {
      debug('found rspack config %s', configFile)
      const sourcedConfig = await importModule(configFile)

      debug('config contains %o', sourcedConfig)
      if (sourcedConfig && typeof sourcedConfig === 'object') {
        userRspackConfig = sourcedConfig.default ?? sourcedConfig
      }
    }

    if (!userRspackConfig) {
      debug('could not find rspack.config!')
      if (config.devServerConfig?.onConfigNotFound) {
        config.devServerConfig.onConfigNotFound('rspack', projectRoot, configFiles)
        // The config process will be killed from the parent, but we want to early exit so we don't get
        // any additional errors related to not having a config
        process.exit(0)
      } else {
        throw new Error(
          `Your Cypress devServer config is missing a required rspackConfig property, since we could not automatically detect one.\n
          Please add one to your ${config.devServerConfig.cypressConfig.configFile}`,
        )
      }
    }
  }

  userRspackConfig =
    typeof userRspackConfig === 'function' ? await userRspackConfig() : userRspackConfig

  const userAndFrameworkRspackConfig = modifyRspackConfigForCypress(
    merge(frameworkRspackConfig ?? {}, userRspackConfig ?? {}),
  )

  debug(
    `User passed in user and framework rspack config with values %o`,
    userAndFrameworkRspackConfig,
  )
  debug(`New rspack entries %o`, files)
  debug(`Project root`, projectRoot)
  debug(`Support file`, supportFile)

  const mergedConfig = merge(userAndFrameworkRspackConfig, makeCypressRspackConfig(config))

  // Some frameworks (like Next.js) change this value which changes the path we would need to use to fetch our spec.
  // (eg, http://localhost:xxxx/<dev-server-public-path>/static/chunks/spec-<x>.js). Deleting this key to normalize
  // the spec URL to `*/spec-<x>.js` which we need to know up-front so we can fetch the sourcemaps.
  delete mergedConfig.output?.chunkFilename

  // Angular loads global styles and polyfills via script injection in the index.html
  if (framework === 'angular') {
    mergedConfig.entry = {
      ...(mergedConfig.entry as EntryObject),
      'cypress-entry': CYPRESS_RSPACK_ENTRYPOINT,
    }
  } else {
    mergedConfig.entry = CYPRESS_RSPACK_ENTRYPOINT
  }

  debug('Merged rspack config %o', mergedConfig)

  return mergedConfig
}
