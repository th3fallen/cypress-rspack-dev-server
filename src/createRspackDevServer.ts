import debugLib from 'debug'
import type { Configuration } from '@rspack/dev-server'
import type { StatsOptions } from 'webpack'
import type { DevServerConfig } from './devServer'
import type { SourceRelativeRspackResult } from './helpers/sourceRelativeRspackModules'
import { makeRspackConfig } from './makeRspackConfig'

const debug = debugLib('cypress-rspack-dev-server:start')

/**
 * Takes the rspack / rspackDevServer modules, the configuration provide
 * from the framework override (if any), and the configuration provided
 * from the user config (if any) and makes the final config we want to
 * serve into rspack
 */
export interface CreateFinalRspackConfig {
  /**
   * Initial config passed to devServer
   */
  devServerConfig: DevServerConfig
  /**
   * Result of sourcing the rspack from the
   */
  sourceRspackModulesResult: SourceRelativeRspackResult
  /**
   * Framework-specific config overrides
   */
  frameworkConfig?: unknown
}

export async function createRspackDevServer(config: CreateFinalRspackConfig) {
  const {
    sourceRspackModulesResult: {
      rspack: { module: rspack },
      rspackDevServer: { module: RspackDevServer },
    },
    devServerConfig: {
      cypressConfig: { devServerPublicPathRoute, isTextTerminal },
    },
  } = config

  const finalRspackConfig = await makeRspackConfig(config)
  const rspackCompiler = rspack(finalRspackConfig)

  const isOpenMode = !isTextTerminal
  const rspackDevServerConfig: Configuration = {
    host: '127.0.0.1',
    port: 'auto',
    ...finalRspackConfig.devServer,
    devMiddleware: {
      publicPath: devServerPublicPathRoute,
      stats: (finalRspackConfig.stats ?? 'minimal') as StatsOptions,
    },
    hot: false,
    // Only enable file watching & reload when executing tests in `open` mode
    liveReload: isOpenMode,
    client: { overlay: false },
  }

  const server = new RspackDevServer(rspackDevServerConfig, rspackCompiler)

  return { server, compiler: rspackCompiler }
}
