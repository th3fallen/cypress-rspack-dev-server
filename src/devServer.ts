import debugLib from 'debug'
import type { EventEmitter } from 'events'
import type { RspackDevServer } from '@rspack/dev-server'
import type { Compiler, Configuration } from '@rspack/core'

import { createRspackDevServer } from './createRspackDevServer'
import {
  sourceDefaultRspackDependencies,
  SourceRelativeRspackResult,
} from './helpers/sourceRelativeRspackModules'

const debug = debugLib('cypress-rspack-dev-server:devServer')

export type Frameworks = Extract<
  Cypress.DevServerConfigOptions,
  { bundler: 'webpack' }
>['framework']

type FrameworkConfig =
  | {
      framework?: Exclude<Frameworks, 'angular'>
    }
  | {
      framework: 'angular'
      options?: {
        projectConfig: Cypress.AngularDevServerProjectConfig
      }
    }

type ConfigHandler =
  | Partial<Configuration>
  | (() => Partial<Configuration> | Promise<Partial<Configuration>>)

export type DevServerConfig = {
  specs: Cypress.Spec[]
  cypressConfig: Cypress.PluginConfigOptions
  devServerEvents: EventEmitter
  onConfigNotFound?: (devServer: 'rspack', cwd: string, lookedIn: string[]) => void
  rspackConfig?: ConfigHandler // Derived from the user's rspack config
} & FrameworkConfig

/**
 * @internal
 */
type DevServerCreateResult = {
  server: RspackDevServer
  compiler: Compiler
}

/**
 * import { RspackDevServer } from '@rspack/dev-server'
 *
 * Creates & returns a RspackDevServer for serving files related
 * to Cypress Component Testing
 *
 * @param config
 */
export function devServer(
  devServerConfig: DevServerConfig,
): Promise<Cypress.ResolvedDevServerConfig> {
  return new Promise(async (resolve, reject) => {
    const result = (await devServer.create(devServerConfig)) as DevServerCreateResult

    result.server
      .start()
      .then(() => {
        if (!result.server.options.port) {
          return reject(new Error(`Expected port ${result.server.options.port} to be a number`))
        }

        debug('Component testing rspack server started on port %s', result.server.options.port)

        resolve({
          port: result.server.options.port as number,
          // Close is for unit testing only. We kill this child process which will handle the closing of the server
          close: async (done) => {
            debug('closing dev server')
            result.server
              .stop()
              .then(() => done?.())
              .catch(done)
              .finally(() => {
                debug('closed dev server')
              })
          },
        })
      })
      .catch(reject)
  })
}

type PresetHandlerResult = {
  frameworkConfig: Configuration
  sourceRspackModulesResult: SourceRelativeRspackResult
}

type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>

const thirdPartyDefinitionPrefixes = {
  // matches @org/cypress-ct-*
  namespacedPrefixRe: /^@.+?\/cypress-ct-.+/,
  globalPrefix: 'cypress-ct-',
}

function isThirdPartyDefinition(framework: string) {
  return (
    framework.startsWith(thirdPartyDefinitionPrefixes.globalPrefix) ||
    thirdPartyDefinitionPrefixes.namespacedPrefixRe.test(framework)
  )
}

async function getPreset(
  devServerConfig: DevServerConfig,
): Promise<Optional<PresetHandlerResult, 'frameworkConfig'>> {
  const defaultRspackModules = () => ({
    sourceRspackModulesResult: sourceDefaultRspackDependencies(devServerConfig),
  })

  // Third party library (eg solid-js, lit, etc)
  if (devServerConfig.framework && isThirdPartyDefinition(devServerConfig.framework)) {
    return defaultRspackModules()
  }

  switch (devServerConfig.framework) {
    // todo - add support for other frameworks
    case 'create-react-app':
    // return createReactAppHandler(devServerConfig);
    case 'nuxt':
    // return await nuxtHandler(devServerConfig)

    case 'vue-cli':
    // return await vueCliHandler(devServerConfig)

    case 'next':
    // return await nextHandler(devServerConfig)

    case 'angular':
    // return await angularHandler(devServerConfig)

    case 'react':
    case 'vue':
    case 'svelte':
    case undefined:
      return defaultRspackModules()

    default:
      throw new Error(
        `Unexpected framework ${
          (devServerConfig as any).framework
        }, please visit https://on.cypress.io/component-framework-configuration to see a list of supported frameworks`,
      )
  }
}

/**
 * Synchronously create the rspack server instance, without starting.
 * Useful for testing
 *
 * @internal
 */
devServer.create = async function (devServerConfig: DevServerConfig) {
  const { frameworkConfig, sourceRspackModulesResult } = await getPreset(devServerConfig)

  const { server, compiler } = await createRspackDevServer({
    devServerConfig,
    frameworkConfig,
    sourceRspackModulesResult,
  })

  return { server, compiler }
}
