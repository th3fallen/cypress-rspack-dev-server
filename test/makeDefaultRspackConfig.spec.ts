import { describe, expect } from '@jest/globals'
import EventEmitter from 'events'
import { CreateFinalRspackConfig } from '../src/createRspackDevServer'
import { makeCypressRspackConfig } from '../src/makeDefaultRspackConfig'
import { createModuleMatrixResult } from './test-helper/createModuleMatrixResult'
import { makeRspackConfig } from '../src/makeRspackConfig'

describe('makeCypressRspackConfig', () => {
  // Returns a valid Configuration object with mode, optimization, output, plugins and devtool properties
  test('should return a valid Configuration object', () => {
    const config: CreateFinalRspackConfig = {
      devServerConfig: {
        cypressConfig: {
          projectRoot: 'path/to/project',
          devServerPublicPathRoute: '/public',
          supportFile: 'path/to/supportFile',
          indexHtmlFile: 'path/to/indexHtmlFile',
          isTextTerminal: true,
        } as Cypress.PluginConfigOptions,
        specs: [],
        devServerEvents: new EventEmitter(),
      },
      sourceRspackModulesResult: createModuleMatrixResult(),
    }

    const result = makeCypressRspackConfig(config)

    expect(result.mode).toBe('development')
    expect(result.optimization).toEqual({
      sideEffects: false,
      splitChunks: { chunks: 'all' },
    })
    expect(result.plugins).toMatchSnapshot()
    expect(result.devtool).toBe('inline-source-map')
    expect(result.optimization?.sideEffects).toBe(false)
  })

  // test('should replace backslashes with posixSeparator in the publicPath when path.sep is not equal to posixSeparator', () => {
  //   const originalPathSep = path.sep
  //   path.sep = '\\'

  //   const config: CreateFinalRspackConfig = {
  //     devServerConfig: {
  //       cypressConfig: {
  //         projectRoot: 'path/to/project',
  //         devServerPublicPathRoute: '/public',
  //         supportFile: 'path/to/supportFile',
  //         indexHtmlFile: 'path/to/indexHtmlFile',
  //         isTextTerminal: true,
  //       } as Cypress.PluginConfigOptions,
  //       specs: ['path/to/spec1', 'path/to/spec2'],
  //       devServerEvents: new EventEmitter(),
  //       framework: 'react',
  //     },
  //     sourceRspackModulesResult: createModuleMatrixResult(),
  //   }

  //   const result = makeCypressRspackConfig(config)

  //   expect(result.output.publicPath).toBe('/public/')
  //   path.sep = originalPathSep
  // })
})

describe('experimentalJustInTimeCompile', () => {
  const devServerConfig: CreateFinalRspackConfig['devServerConfig'] = {
    specs: [],
    cypressConfig: {
      projectRoot: '.',
      indexHtmlFile: 'path/to/supportFile',
      devServerPublicPathRoute: '/test-public-path',
      justInTimeCompile: true,
      baseUrl: null,
    } as Cypress.PluginConfigOptions,
    rspackConfig: {
      entry: { main: 'src/index.js' },
    },
    devServerEvents: new EventEmitter(),
  }

  describe('run mode', () => {
    beforeEach(() => {
      devServerConfig.cypressConfig.isTextTerminal = true
    })

    it('enables watching', async () => {
      const actual = await makeRspackConfig({
        devServerConfig,
        sourceRspackModulesResult: createModuleMatrixResult(),
      })

      expect(actual.watchOptions?.ignored).toStrictEqual(/node_modules/)
    })
  })
})
