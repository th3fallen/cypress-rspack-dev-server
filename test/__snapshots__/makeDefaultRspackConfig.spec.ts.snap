// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeCypressRspackConfig should return a valid Configuration object 1`] = `
[
  HtmlRspackPlugin {
    "_options": {
      "filename": "index.html",
      "inject": "head",
      "meta": {},
      "scriptLoading": "defer",
      "template": "path/to/indexHtmlFile",
    },
    "affectedHooks": undefined,
    "name": "HtmlRspackPlugin",
  },
  CypressCTRspackPlugin {
    "addCompilationHooks": [Function],
    "addLoaderContext": [Function],
    "beforeCompile": [Function],
    "compilation": null,
    "devServerEvents": EventEmitter {
      "_events": {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
      Symbol(kCapture): false,
    },
    "files": [],
    "indexHtmlFile": "path/to/indexHtmlFile",
    "onSpecsChange": [Function],
    "projectRoot": "path/to/project",
    "rspack": {
      "RspackDevServer": [Function],
    },
    "supportFile": "path/to/supportFile",
  },
]
`;
